src/librustc_mir/build/matches/mod.rs

```diff
@@ -512,7 +512,10 @@ impl<'a,'tcx> Builder<'a,'tcx> {
         let otherwise: Vec<_> =
             target_blocks.into_iter()
                          .zip(target_candidates)
-                         .flat_map(|(target_block, target_candidates)| {
+                         .flat_map(|(target_block, mut target_candidates)| {
+                             // We need to preserve the fact that the candidates
+                             // are in the reversed order compared to the source.
+                             target_candidates.reverse();
                              self.match_candidates(span,
```

Comment:
Perhaps `match_candidates` could be refactored to consume an iterator, thus allowing us to avoid a potentially expensive operation that is reversal of elements in a `Vec`tor?